### ISSUE TITLE
[Feature Requests] Logging and authentication

### ISSUE THREAD

**Issue by Sidefix** [halverneus/static-file-server#26]
Hello,

I have been using this file server for a bit, and I love it; I have noted below one semi-critical request, and one long-term request:

For context, I am running this server on a Linux machine.
- Logging is only semi-useful and does not appear to echo correctly. To elaborate: when the server is running and is accessed from an external address, I would expect logging to poll the requester IP. Instead, both addresses seen are for the host. As an example, two IP addresses: 10.20.30.40 (host/server) and 40.30.20.10 (client/requester). 
I would expect the logs to read: "REQ from '40.30.20.10': GET HTTP/1.1 10.20.30.40/Content/Img.jpg", however the current read is: "REQ from '10.20.30.40': GET HTTP/1.1 10.20.30.40/Content/Img.jpg"; i.e. the source requester is never logged.
The second issue with logging is that if I try to echo the output in a file, the above never gets logged. All I ever see is the initial server config message.

- HTTP user authentication. Long term request, as I imagine how complicated this would need to be to work, but it would be amazing if this could be achieved.

Thank you in advance for your consideration :)

**Comment by halverneus** [comment]
- Looks like code is using the REFERER rather than the remote address in the "REQ from ???". I'll make a change for it to use the remote address, but also include the Referrer information as part of the merge request for adding Referrer filtering.
- As for the logging, the configuration output is being printed to stdout and the later logging is printed to stderr ("serve > out.log" will send the configuration and "serve 2> out.log" will send request logging and "serve > out.log 2>&1" will send both). This was initially intentional so that a user had a choice to send logging output without betraying their configuration, but I don't have a specific use-case for that demand. What are your thoughts? (keep as is, send both to stdout or send both to stderr (or other?))
- I did create a branch a while back for Basic Authentication, but it didn't get much attention. Having said that, check out the help documentation located at the top of [this file](https://github.com/halverneus/static-file-server/blob/feature/basic_auth/serve.go) and leave me some feedback on [this issue](https://github.com/halverneus/static-file-server/issues/5) and I'll see about adding it in. Mostly, I just want to make sure the solution is secure, simple and something users actually want (which makes your feedback extremely valuable!).

**Comment by halverneus** [comment]
As a chance to provide some feedback on logging, this is how logging will change:

Originally:
```
2019/04/08 11:54:05 REQ from 'http://localhost': GET HTTP/1.1 localhost:8080/main.go -> bin/serve/main.go
```

After the change (without and with the REFERER header being set):
```
2019/04/08 12:19:41 REQ from '127.0.0.1:52718': GET HTTP/1.1 localhost:8080/main.go -> bin/serve/main.go
2019/04/08 12:21:20 REQ from '127.0.0.1:52750' (REFERER: 'http://localhost'): GET HTTP/1.1 localhost:8080/main.go -> bin/serve/main.go
```

**Comment by Sidefix** [comment]
@halverneus hey, thanks for your reply!

Regarding the logging, I never considered that the output (apart from the config) was being logged as err... lack of lateral thinking on my part :)
I do believe that it would be more efficient to log everything as stdout, to avoid confusion; or, at the very least, document this behavior (unless it already was and I had just missed it).

Regarding the logging format, yes the new format proposed above with the referer header goes a long way to providing adequate access information; much appreciated!

Regarding authentication, I don't see why someone wouldn't want authenticated access to a file server, as long as its configurable; i.e. still have the option of anonymous access if one so desires. And just to nip this in the bud ahead of time, this could be part of the logging as well (i.e. add the username to the remote address in the logs for clarity). 

Although I am not familiar with GO, I noted a few things:

- The credentials would be stored in a JSON
This is great, because it allows offline or even granular or automatic administration of users

- I'm a bit sketchy on the usefulness of FAST AUTH, maybe because the scope of the feature is rather vague

- Immediate suggestion would be to add a whitelist feature for IP addresses as an alternative for credentials or anonymous access

I will provide this feedback on the original authentication open ticket.

Thank you very much for being involved in this!
Best regards

**Comment by halverneus** [comment]
Thanks! I'll close this as soon as I release the changes (probably the last half of this week).

**Comment by halverneus** [comment]
Logging updates are now available in v1.6.4. Authentication features are continuing to be tracked in #5 .
